server/ocsp.go

```diff
@@ -0,0 +1,442 @@
+	key := fmt.Sprintf("%x", sha256.Sum256(oc.Leaf.Raw))
+	opts := oc.srv.getOpts()
+	stateDir := opts.StateDir
+	oc.mu.Lock()
+	raw, err := ioutil.ReadFile(filepath.Join(stateDir, key))
+	oc.mu.Unlock()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resp, err := ocsp.ParseResponse(raw, oc.Issuer)
+	if err != nil {
+		return nil, nil, err
+	}
+	if err := validOCSPResponse(resp); err != nil {
+		return nil, nil, err
+	}
+
+	oc.mu.Lock()
+	oc.raw = raw
+	oc.resp = resp
+	oc.mu.Unlock()
+
+	return raw, resp, nil
+}
+
+func (oc *OCSPMonitor) getRemoteStatus() ([]byte, *ocsp.Response, error) {
+	opts := oc.srv.getOpts()
+	stateDir := opts.StateDir
+	overrideURLs := opts.OCSPConfig.OverrideURLs
+	getRequestBytes := func(u string, hc *http.Client) ([]byte, error) {
+		resp, err := hc.Get(u)
+		if err != nil {
+			return nil, err
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("non-ok http status: %d", resp.StatusCode)
+		}
+
+		return ioutil.ReadAll(resp.Body)
+	}
+
+	// Request documentation:
+	// https://tools.ietf.org/html/rfc6960#appendix-A.1
+
+	reqDER, err := ocsp.CreateRequest(oc.Leaf, oc.Issuer, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	reqEnc := base64.StdEncoding.EncodeToString(reqDER)
+
+	responders := oc.Leaf.OCSPServer
+	if len(overrideURLs) > 0 {
+		responders = overrideURLs
+	}
+
+	oc.mu.Lock()
+	hc := oc.hc
+	oc.mu.Unlock()
+	var raw []byte
+	for _, u := range responders {
+		u = strings.TrimSuffix(u, "/")
+		raw, err = getRequestBytes(fmt.Sprintf("%s/%s", u, reqEnc), hc)
+		if err == nil {
+			break
+		}
+	}
+	if err != nil {
+		return nil, nil, fmt.Errorf("exhausted ocsp servers: %w", err)
+	}
+
+	resp, err := ocsp.ParseResponse(raw, oc.Issuer)
+	if err != nil {
+		return nil, nil, err
+	}
+	if err := validOCSPResponse(resp); err != nil {
+		return nil, nil, err
+	}
+
+	key := fmt.Sprintf("%x", sha256.Sum256(oc.Leaf.Raw))
+	if err := oc.writeOCSPStatus(stateDir, key, raw); err != nil {
+		return nil, nil, fmt.Errorf("failed to write ocsp status: %w", err)
+	}
+
+	oc.mu.Lock()
+	oc.raw = raw
+	oc.resp = resp
+	oc.mu.Unlock()
+
+	return raw, resp, nil
+}
+
+func (oc *OCSPMonitor) run() {
+	s := oc.srv
+	defer s.grWG.Done()
+	config := s.getOpts().OCSPConfig
+	shouldShutdown := config != nil && (config.Mode == OCSPModeMust || config.Mode == OCSPModeAlways)
+
+	var nextRun time.Duration
+	_, resp, err := oc.getStatus()
+	if err == nil && resp.Status == ocsp.Good {
+		nextRun = oc.getNextRun()
+		t := resp.NextUpdate.Format(time.RFC3339Nano)
+		s.Noticef(
+			"Found existing OCSP certificate status: good, next update %s, checking again in %s",
+			t, nextRun,
+		)
+	} else if err == nil && shouldShutdown {
+		// If resp.Status is ocsp.Revoked, ocsp.Unknown, or any other value.
+		s.Errorf("Found existing OCSP certificate status: %s", ocspStatusString(resp.Status))
+		s.Shutdown()
+		return
+	}
+
+	for {
+		time.Sleep(nextRun)
```

Comment:
These make me nervous, should be a select and timeout etc.

Reply:
Updated to use a select and also using the quitCh so that it does not block.